File: packages/backend/sdk/src/resources/accounts.ts
import type { FluxisClient } from '../client.js';
import type {
  Account,
  AccountSettlementAddresses,
  CreateAccountRequest,
  SettlementAddressRequest,
  UpdateAccountRequest,
} from '../types/accounts.js';
import type { SettlementAddressResponse } from '../types/organization.js';

/** Canonical read path on core-api (`GET /account/{accountId}/settlement-addresses`). */
function accountSettlementReadPath(accountId: string): string {
  return `/account/${accountId}/settlement-addresses`;
}

/**
 * Write path on core-api. POST/PUT/DELETE (and a redundant GET) live under
 * `/account/settlement/{accountID}/settlement-addresses`.
 */
function accountSettlementWritePath(accountId: string): string {
  return `/account/settlement/${accountId}/settlement-addresses`;
}

export class AccountsResource {
  constructor(private readonly client: FluxisClient) {}

  async list(): Promise<Account[]> {
    return this.client.request<Account[]>('GET', '/account');
  }

  async get(accountId: string): Promise<Account> {
    return this.client.request<Account>('GET', `/account/${accountId}`);
  }

  async create(data: CreateAccountRequest): Promise<Account> {
    return this.client.request<Account>('POST', '/account', data);
  }

  async update(accountId: string, data: UpdateAccountRequest): Promise<Account> {
    return this.client.request<Account>('PUT', `/account/${accountId}`, data);
  }

  async delete(accountId: string): Promise<void> {
    await this.client.request<void>('DELETE', `/account/${accountId}`);
  }

  async getSettlementAddresses(accountId: string): Promise<AccountSettlementAddresses> {
    return this.client.request<AccountSettlementAddresses>(
      'GET',
      accountSettlementReadPath(accountId),
    );
  }

  async setSettlementAddress(
    accountId: string,
    data: SettlementAddressRequest,
  ): Promise<SettlementAddressResponse> {
    return this.client.request<SettlementAddressResponse>(
      'POST',
      accountSettlementWritePath(accountId),
      data,
    );
  }

  async updateSettlementAddress(
    accountId: string,
    data: SettlementAddressRequest,
  ): Promise<SettlementAddressResponse> {
    return this.client.request<SettlementAddressResponse>(
      'PUT',
      accountSettlementWritePath(accountId),
      data,
    );
  }

  async deleteSettlementAddress(accountId: string, network: string): Promise<void> {
    await this.client.request<void>(
      'DELETE',
      accountSettlementWritePath(accountId),
      undefined,
      { network },
    );
  }
}

File: packages/backend/sdk-python/src/fluxis/resources/accounts.py
"""Account resource."""

from __future__ import annotations

from typing import TYPE_CHECKING

from fluxis.models.account import (
    Account,
    AccountSettlementAddresses,
    CreateAccountRequest,
    SettlementAddressRequest,
    SettlementAddressResponse,
    UpdateAccountRequest,
)
from fluxis.models.common import parse_list

if TYPE_CHECKING:
    from fluxis.client import AsyncFluxisClient, FluxisClient


class AccountsResource:
    def __init__(self, client: FluxisClient) -> None:
        self._client = client

    def list(self) -> list[Account]:
        data = self._client.request("GET", "/account")
        return parse_list(data, Account)

    def get(self, account_id: str) -> Account:
        data = self._client.request("GET", f"/account/{account_id}")
        return Account.from_dict(data)

    def create(self, data: CreateAccountRequest) -> Account:
        result = self._client.request("POST", "/account", data)
        return Account.from_dict(result)

    def update(self, account_id: str, data: UpdateAccountRequest) -> Account:
        result = self._client.request("PUT", f"/account/{account_id}", data)
        return Account.from_dict(result)

    def delete(self, account_id: str) -> None:
        self._client.request("DELETE", f"/account/{account_id}")

    def get_settlement_addresses(self, account_id: str) -> AccountSettlementAddresses:
        data = self._client.request("GET", f"/account/{account_id}/settlement-addresses")
        return AccountSettlementAddresses.from_dict(data)

    def set_settlement_address(
        self,
        account_id: str,
        data: SettlementAddressRequest,
    ) -> SettlementAddressResponse:
        result = self._client.request("POST", f"/account/{account_id}/settlement-addresses", data)
        return SettlementAddressResponse.from_dict(result)

    def update_settlement_address(
        self,
        account_id: str,
        data: SettlementAddressRequest,
    ) -> SettlementAddressResponse:
        result = self._client.request("PUT", f"/account/{account_id}/settlement-addresses", data)
        return SettlementAddressResponse.from_dict(result)

    def delete_settlement_address(self, account_id: str, network: str) -> None:
        self._client.request(
            "DELETE",
            f"/account/{account_id}/settlement-addresses",
            query={"network": network},
        )


class AsyncAccountsResource:
    def __init__(self, client: AsyncFluxisClient) -> None:
        self._client = client

    async def list(self) -> list[Account]:
        data = await self._client.request("GET", "/account")
        return parse_list(data, Account)

    async def get(self, account_id: str) -> Account:
        data = await self._client.request("GET", f"/account/{account_id}")
        return Account.from_dict(data)

    async def create(self, data: CreateAccountRequest) -> Account:
        result = await self._client.request("POST", "/account", data)
        return Account.from_dict(result)

    async def update(self, account_id: str, data: UpdateAccountRequest) -> Account:
        result = await self._client.request("PUT", f"/account/{account_id}", data)
        return Account.from_dict(result)

    async def delete(self, account_id: str) -> None:
        await self._client.request("DELETE", f"/account/{account_id}")

    async def get_settlement_addresses(self, account_id: str) -> AccountSettlementAddresses:
        data = await self._client.request("GET", f"/account/{account_id}/settlement-addresses")
        return AccountSettlementAddresses.from_dict(data)

    async def set_settlement_address(
        self,
        account_id: str,
        data: SettlementAddressRequest,
    ) -> SettlementAddressResponse:
        result = await self._client.request(
            "POST",
            f"/account/{account_id}/settlement-addresses",
            data,
        )
        return SettlementAddressResponse.from_dict(result)

    async def update_settlement_address(
        self,
        account_id: str,
        data: SettlementAddressRequest,
    ) -> SettlementAddressResponse:
        result = await self._client.request(
            "PUT",
            f"/account/{account_id}/settlement-addresses",
            data,
        )
        return SettlementAddressResponse.from_dict(result)

    async def delete_settlement_address(self, account_id: str, network: str) -> None:
        await self._client.request(
            "DELETE",
            f"/account/{account_id}/settlement-addresses",
            query={"network": network},
        )
